T5: Arrays and collections

denyaembla opened the pull request.

denyaembla committed:
T5: Reassembled task, added Menu class.

denyaembla committed:
T5: Added a few Display methods, fixed GroceryBag error.

denyaembla committed:
T5: Reworked Menu, added method that aggregates Menu methods

denyaembla committed:
T5: Added red-color to Alcohol validation method. Fixed MainMenu collection-related errors.

denyaembla committed:
T5: Added sln., csproj. and Program.cs files

DaryaNikitina commented on TASK-5/TASK-5/Generators/ItemGenerator.cs:
```diff
@@ -0,0 +1,54 @@
+    private static readonly int BarcodeMaximumValue = 9999;
+    private static readonly decimal MinimumPrice = 0.7M;
+    private static readonly decimal MaximumPrice = 5M;
+    private static readonly string[] ItemNames = {"juice", "milk", "ice-cream", "bread", "** BEER **"};
```
Use Bogus instead of hardcoding product names.

DaryaNikitina replied:
Use Bogus instead of hardcoding product names.

DaryaNikitina commented on TASK-5/TASK-5/Generators/ItemGenerator.cs:
```diff
@@ -0,0 +1,54 @@
+    
+    public static Item GenerateItem()
+    {
+        var price = new Faker().Random.Decimal(MinimumPrice, MaximumPrice);
+        price = decimal.Round(price, 2);
```
Remove and use inside Item creation

DaryaNikitina replied:
Remove and use inside Item creation

DaryaNikitina commented on TASK-5/TASK-5/Generators/ItemGenerator.cs:
```diff
@@ -0,0 +1,54 @@
+        return item.Generate();
+    }
+
+    internal static List<Item> GenerateItems()
+    {
+        var itemBag = new List<Item>();
+        for (var i = 0; i < 4; i++)
+        {
+            itemBag.Add(GenerateItem());
+        }
+
+        return itemBag;
+    }
```
Make it simple. Generate method can pass count.
Ex, item.Generate(count) - for method above

DaryaNikitina replied:
Make it simple. Generate method can pass count.
Ex, item.Generate(count) - for method above

DaryaNikitina commented on TASK-5/TASK-5/Models/Item.cs:
```diff
@@ -0,0 +1,9 @@
+public class Item
+{
+    public int Barcode { get; set; }
+    public string? ItemName { get; set; }
+    public string? Category { get; set; }
```
?

DaryaNikitina replied:
?

DaryaNikitina commented on TASK-5/TASK-5/Models/User.cs:
```diff
@@ -0,0 +1,9 @@
+    public int PassportId { get; set; }
+    public string FullName { get; set; }
+    public int Age { get; set; }
+    public List<Item>? GroceryBag { get; set; }
```
?

DaryaNikitina replied:
?

DaryaNikitina commented on TASK-5/TASK-5/Program.cs:
```diff
@@ -0,0 +1,17 @@
+﻿// See https://aka.ms/new-console-template for more information
+
+
```
Remove.

DaryaNikitina replied:
Remove.

DaryaNikitina commented on TASK-5/TASK-5/Validation.cs:
```diff
@@ -0,0 +1,34 @@
+{
+    public static bool SameIdChecker(int passportId, List<User> users)
+    {
+        foreach (var u in users)
```
What's naming for element in users "u"? You could name it as user.

DaryaNikitina replied:
What's naming for element in users "u"? You could name it as user.

DaryaNikitina commented on TASK-5/TASK-5/Program.cs:
```diff
@@ -0,0 +1,17 @@
+
+var result = Menu.GenerateFiveUsers();
+Menu.MainMenu(result);
+
+
+
+
+
+
+
+
```
Spaces

DaryaNikitina replied:
Spaces

DaryaNikitina commented on TASK-5/TASK-5/Menu.cs:
```diff
@@ -0,0 +1,176 @@
+        Console.WriteLine($"Total: {totalPrice}$");
+    }
+   
+    private static void DisplayItems(List<Item> itemBag)     //why am i not using this???
```
Idk, use it.

DaryaNikitina replied:
You can use it in DisplayClientPurchases method

DaryaNikitina commented on TASK-5/TASK-5/Menu.cs:
```diff
@@ -0,0 +1,176 @@
+        Console.WriteLine($"Total: {totalPrice}$");
+    }
+   
+    private static void DisplayItems(List<Item> itemBag)     //why am i not using this???
```
Idk, use it.

DaryaNikitina commented on TASK-5/TASK-5/Menu.cs:
```diff
@@ -0,0 +1,176 @@
+        Console.WriteLine($"Total: {totalPrice}$");
+    }
+    
+    private static void DisplayClientPurchases(List<User> users)            //overload with manual number input
```
Remove comment.

DaryaNikitina replied:
Remove comment.

DaryaNikitina commented on TASK-5/TASK-5/Menu.cs:
```diff
@@ -0,0 +1,176 @@
+        }
+    }
+
+    private static void DisplayClientPurchases(List<User> users, int userNumber)        //counting from 1;
```
Remove comment.

DaryaNikitina replied:
Remove comment.

DaryaNikitina commented on TASK-5/TASK-5/Menu.cs:
```diff
@@ -0,0 +1,176 @@
+        }
+    }
+
+    public static List<User> GenerateFiveUsers()
```
Make it GenerateUsers and pass count parameter.

DaryaNikitina commented on TASK-5/TASK-5/Menu.cs:
```diff
@@ -0,0 +1,176 @@
+        }
+    }
+
+    public static List<User> GenerateFiveUsers()
```
Make it GenerateUsers and pass count parameter.

DaryaNikitina replied:
UserGenerator.GenerateUserWithItems(int count) -> userWithItems.Generate(int count)

DaryaNikitina commented on TASK-5/TASK-5/Menu.cs:
```diff
@@ -0,0 +1,176 @@
+        }
+    }
+    
+    private static void GenerateUserFromConsole(List<User> users) 
+    {
+        Console.Write("You are going to create new user \n Enter new user's passportID \n");
+        var inputPassportId = Convert.ToInt32(Console.ReadLine());
+
+        if (Validation.SameIdChecker(inputPassportId, users))
+        {
+            users.Add(UserGenerator.GenerateUserFromConsole(inputPassportId));
+            Console.WriteLine("You've added new user successfully");
+        }
+        else
+        {
+            Console.WriteLine("User cannot be added");
+        }
+    }
```
Why do you have similar names for methods?
Menu.GenerateUserFromConsole(inputPassportId)
UserGenerator.GenerateUserFromConsole(inputPassportId)

You can name it as CreateUser

DaryaNikitina replied:
Why do you have similar names for methods?
Menu.GenerateUserFromConsole(inputPassportId)
UserGenerator.GenerateUserFromConsole(inputPassportId)

You can name it as CreateUser

DaryaNikitina commented on TASK-5/TASK-5/Menu.cs:
```diff
@@ -0,0 +1,176 @@
+
+public class Menu
+{
+    private static readonly Random _amount = new();
```
Remove it

DaryaNikitina replied:
Remove it

DaryaNikitina commented on TASK-5/TASK-5/Menu.cs:
```diff
@@ -0,0 +1,176 @@
+            Console.WriteLine("User cannot be added");
+        }
+    }
+
+    
+
```
Spaces.

DaryaNikitina replied:
Spaces.

DaryaNikitina commented on TASK-5/TASK-5/Generators/UserGenerator.cs:
```diff
@@ -0,0 +1,53 @@
+                faker => new User())
+            .RuleFor(u => u.FullName, (f, u)=> f.Name.FullName())
+            .RuleFor(u => u.Age, (f, u) => f.Random.Int(MinimumAge, MaximumAge))
+            .RuleFor(u => u.PassportId, (f, u)=> _counter++)
```
Make passportId more dynamic. Random digits with given length.

DaryaNikitina replied:
Make passportId more dynamic. Random digits with given length.

denyaembla committed:
T5: Reworked according to comments:
Changed ID type (and ID Validation accordingly);
Changed names of some methods;
Added validation for GenerateFromConsole method.

DaryaNikitina commented on TASK-5/TASK-5/Generators/ItemGenerator.cs:
```diff
@@ -0,0 +1,50 @@
+        }
+
+        return itemBag;
+    }
```
Delete this method and for GenerateItem pass count, make GenerateItem return value as List<Item>

DaryaNikitina commented on TASK-5/TASK-5/Menu.cs:
```diff
@@ -0,0 +1,172 @@
+        Console.WriteLine("Enter new user's ID");
+        var userPassportIdInput = Console.ReadLine();
+        Guid passportId;
+        while (string.IsNullOrWhiteSpace(userPassportIdInput) || userPassportIdInput.Length != 32 
+                                                              || !Guid.TryParse(userPassportIdInput, out passportId))
+        {
+            Console.WriteLine("User's passport ID cannot be empty and it's length has to be 32 symbols");
+            userPassportIdInput = Console.ReadLine();
+        }
```
Make validation method with regex for passportId(In Validation class, for example)

DaryaNikitina commented on TASK-5/TASK-5/Generators/UserGenerator.cs:
```diff
@@ -0,0 +1,73 @@
+    {
+        Console.Write("Enter new user's name \n");
+        var inputName = Console.ReadLine();
+        while (string.IsNullOrWhiteSpace(inputName) || inputName.Any(char.IsDigit))
+        {
+            Console.WriteLine("User's firstname cannot contain numbers or be empty.");
+            inputName = Console.ReadLine();
+        }
+        
+        Console.Write("Enter new user's lastname \n");
+        var inputLastname = Console.ReadLine();
+        while (string.IsNullOrWhiteSpace(inputLastname) || inputLastname.Any(char.IsDigit))
+        {
+            Console.WriteLine("User's lastname cannot contain numbers or be empty.");
+            inputLastname = Console.ReadLine();
+        }
```
Make validation method with regex for input name, lastname(In Validation class, for example)

DaryaNikitina commented on TASK-5/TASK-5/Generators/UserGenerator.cs:
```diff
@@ -0,0 +1,73 @@
+        Console.Write("Enter new user's age \n");
+        var inputAge = Console.ReadLine();
+        int age;
+        while (inputAge == null || !int.TryParse(inputAge, out age))
+        {
+            Console.WriteLine("User's age cannot contain letters or be empty.");
+            inputAge = Console.ReadLine();
+        }
```
Make validation method with regex for age(In Validation class, for example)

DaryaNikitina commented on TASK-5/TASK-5/Menu.cs:
```diff
@@ -0,0 +1,172 @@
+        Console.WriteLine("You are in a process of generating item:\n" +
+                          " Please, enter 1 to add a non-alcohol item, 2 for an alcohol item only");
+        var temporaryItem = new Item();
+        switch (Convert.ToInt32(Console.ReadLine()))
+        {
+            case 1:
+                temporaryItem = ItemGenerator.GenerateItem();
+                break;
+            case 2:
+                temporaryItem = ItemGenerator.GenerateAlcoholItem();
+                break;
+        }
```
Create a case for manual creation of item(input Barcode, ItemName, Category, Price)

denyaembla committed:
TASK-5: Added RegEx validations for User.cs fields. Changed Item generation method according to comments.

denyaembla committed:
TASK-5: Committed Validation.cs

DaryaNikitina commented on TASK-5/TASK-5/Validation.cs:
```diff
@@ -0,0 +1,75 @@
+
+            return true;
+        }
+        
```
Empty space

DaryaNikitina commented on TASK-5/TASK-5/Validation.cs:
```diff
@@ -0,0 +1,75 @@
+        private const string OnlyLettersPattern = "^[a-zA-Z]+$";
+        private const string GuidPattern = @"\w{32}$";
+        private const string AgePattern = @"^[1-9]{2}$";
+        
```
Empty space

DaryaNikitina replied:
Empty space

DaryaNikitina commented on TASK-5/TASK-5/Validation.cs:
```diff
@@ -0,0 +1,75 @@
+
+            return true;
+        }
+        public static bool GuidValidation(string input)
```
Add space before method

DaryaNikitina replied:
Add space before method

denyaembla committed:
TASK-5: Code cleanup
